criu/pie/parasite.c

```diff
@@ -211,6 +211,42 @@ static int dump_thread_common(struct parasite_dump_thread *ti)
 	return ret;
 }
 
+/*
+ * Returns a membarrier() registration command (it is a bitmask) if the process
+ * was registered for specified (as a bit index) membarrier()-issuing command;
+ * returns zero otherwise.
+ */
+static int get_membarrier_registration_mask(int cmd_bit)
+{
+	unsigned cmd = 1 << cmd_bit;
+	int ret;
+
+	/*
+	 * Issuing a barrier will be successful only if the process was registered
+	 * for this type of membarrier. All errors are a sign that the type issued
+	 * was not registered (EPERM) or not supported by kernel (EINVAL or ENOSYS).
+	 */
+	ret = sys_membarrier(cmd, 0, 0);
+	if (ret && ret != EPERM && ret != -EINVAL && ret != -ENOSYS) {
+		pr_err("membarrier(1 << %d) returned %d\n", cmd_bit, ret);
+		return ret;
```

Comment:
Maybe `return -1` ? I don't see any reason to return `-E<smth>`. It is normally a bad sign if we start returning something different from -1 on error paths except when we need it.

Also consider printing strerror(-ret) instead of just ret in the above pr_err.

Reply:
The dump_misc() function already returns the -errno from a syscall (prctl). This is the parasite code - I see no strerror() uses here.

Reply:
ret != -EPERM ?

Reply:
> The dump_misc() function already returns the -errno from a syscall (prctl). 

This is the bad argument. It doesn't matter if any other code do the same thing. We should have reasons to do any specific actions. In error cases, we prefer to return -1, because it is very deterministic and hard to handle it incorrectly.